### DISCUSSION TITLE
Ownership analysis for deterministic GC/deallocation

### DISCUSSION THREAD

**Discussion by fabianoliver** [dotnet/runtime#97292] [Ideas]
Apologies if this idea/question is rather high-level and not very specific. I've been curious though if the following would be possible and sensible at all though.

_If you're about to think "well someone just read about Rust's memory ownership model and spend all of 5 minutes thinking whether .NET could benefit from something very vaguely similar" - then you'd be absolutely spot on!_

Right then, say some ref-type is allocated. Would it be possible to attempt figuring out lifetime ownership of the object, and - if a clear ownership can be determined - exclude this allocation from regular GC to instead de-allocate deterministically once (implicitly) released by the owner?

Let's pencil out an example for clarity:

------

### Example 1

```
void example()
{
  var obj = new object();
  DoSomethingWith(obj);
  Console.WriteLine(obj);
}
```

When `obj` is allocated, we can assume the local scope of `example` has ownership of the object.

When `DoSomethingWith`is called, it'd take an analysis of the actual implementation of DoSomethingWith() to figure out what happens to ownership here.
1. If the object reference does not escape the implementation of `DoSomethingWith()`, we can assume obj is passed as a borrowed ref. Which in turn means, once the call concludes, ownership transfers back to the caller - and in this case, `example()` could release the object at the end of the function call
2. If DoSomethingWith does actually store the object reference anywhere that outlasts the actual call, the reference effectively becomes shared - and we leave it up to regular GC
3. If `example` doesn't use `obj` anymore after the call to `DoSomethingWith`  (e.g. we comment out the `Console.WriteLine`-part) , we could assume that `DoSomethingWith` takes full ownership of `obj`. What exactly that means in terms of deallocation would similarly depend on an analysis of `DoSomethingWith` 

This probably gets trickier once async/await is involved; presumably, in case 1, I imagine you could still consider the reference borrowed if `DoSomethingWith` was async, but the call is awaited.

### Example 2

```
class Example {
  private object? _state = null;

  void DoSomethingWith(object obj)
  {
    _state = obj;
  }
}
```

This is probably a whole lot harder in terms of analysis, but in theory - if `DoSomethingWith` in this example is always called with ownership of `obj`, then `_state = obj` could potentially de-allocate the prior value of `_state` when called.
Added complexity of course is that it might _still_ need to be GC'ed in case the instance of `Example` itself is deallocated/collected.

---------

In a purely internal call chain, I'd assume this type of analysis could maybe be done at compile time; for everything else, I'd guess it'd be more of a JIT-task.

While it's rather hard to guess what the practical impact would - anecdotally, a very significant allocations in the codebases I personally tend to work with come from call chains in which _in theory_ there would seem to be a relatively clear path/transfer of ownership.
Think of your classic ASP.NET or GRPC service: Some service receives a web/grpc request, fetches a bunch of data, does a bunch of transformations, funnels it back into an http or grpc response - I think most of that does not entail any form of shared ownership, so I suspect the vast majority of allocations created in such a pipeline _could_ possibly be de-allocated deterministically with sufficiently good flow analysis.

Of course, it's an extremely.. not very fleshed out idea, so I'd be curious - are there any theoretical reasons why this _wouldn't_ work? If not, how feasible or unfeasible would this be in practice?

There have been plenty of prior [prior debate](https://github.com/dotnet/runtime/blob/main/docs/design/coreclr/jit/object-stack-allocation.md) around escape analysis in order to avoid [heap allocations](https://github.com/dotnet/runtime/issues/4584) - there's probably a degree of overlap here; though die idea above would aim more towards (some) deterministic deallocations in general, rather than heap vs. stack allocations specifically. Though I'd imagine ownership analysis would likely be a superset of escape analysis, and thereby could probably _also_ used to stack-allocate some reference types in these scenarios.

**Comment by huoyaoyuan** [comment]
#1661 and #11192

@dotnet/gc is there any recent interest on this?

**Reply by Maoni0** [reply]
as OP said this is a JIT task. the main issue is here: https://github.com/dotnet/runtime/issues/11192
